File: Scripts/statistical_corpus_runner.py
"""Execute the shared statistical corpus through the Python and TypeScript engines."""

from __future__ import annotations

import json
import os
import shutil
import subprocess
import sys
from collections.abc import Callable
from pathlib import Path
from typing import Any

from backend.simulation_models import SimulationCommand, SimulationResult
from backend.simulation_service import run_simulation
from backend.simulation_value_objects import SimulationSeed

ROOT = Path(__file__).resolve().parents[1]
TYPESCRIPT_BRIDGE = ROOT / "frontend/scripts/run-statistical-reference-corpus.mjs"

CanonicalResult = dict[str, Any]
CaseExecutor = Callable[[dict[str, Any]], CanonicalResult]


def canonicalize_python_result(result: SimulationResult) -> CanonicalResult:
    """Rename language-specific fields without rounding, sorting, or filling absent values."""

    canonical: CanonicalResult = {
        "result_kind": result.result_kind,
        "result_percentiles": result.result_percentiles.to_dict(),
    }
    if result.risk_score is not None:
        canonical["risk_score"] = result.risk_score
    canonical["result_distribution"] = [
        {"x": bucket.x, "count": bucket.count} for bucket in result.result_distribution
    ]
    if result.completion_summary is not None:
        summary = result.completion_summary
        canonical["completion_summary"] = {
            "completed_count": summary.completed_count,
            "censored_count": summary.censored_count,
            "censored_rate": summary.censored_rate,
            "horizon_weeks": summary.horizon_weeks,
        }
    canonical["samples_count"] = result.samples_count
    reliability = result.throughput_reliability
    canonical["throughput_reliability"] = {
        "cv": reliability.cv,
        "iqr_ratio": reliability.iqr_ratio,
        "slope_norm": reliability.slope_norm,
        "label": reliability.label,
        "samples_count": reliability.samples_count,
    }
    canonical["seed"] = result.seed.value
    return canonical


def execute_python_case(reference_case: dict[str, Any]) -> CanonicalResult:
    input_value = reference_case["input"]
    command = SimulationCommand.create(
        throughput_samples=input_value["throughput_samples"],
        include_zero_weeks=input_value["include_zero_weeks"],
        mode=input_value["mode"],
        backlog_size=input_value.get("backlog_size"),
        target_weeks=input_value.get("target_weeks"),
        n_sims=input_value["n_sims"],
        seed=SimulationSeed(reference_case["seed"]),
    )
    return canonicalize_python_result(run_simulation(command))


def error_payload(error: object) -> dict[str, str]:
    if isinstance(error, BaseException):
        return {"type": type(error).__name__, "message": str(error)}
    return {"type": type(error).__name__, "message": str(error)}


def _engine_header(engine: str, corpus: dict[str, Any]) -> dict[str, Any]:
    return {
        "engine": engine,
        "corpus_id": corpus["corpus_id"],
        "schema_version": corpus["schema_version"],
        "prng_contract": corpus["prng_contract"]["id"],
    }


def run_python_corpus(
    corpus: dict[str, Any],
    execute_case: CaseExecutor = execute_python_case,
) -> dict[str, Any]:
    case_reports: list[dict[str, Any]] = []
    for reference_case in corpus["cases"]:
        try:
            case_reports.append(
                {
                    "id": reference_case["id"],
                    "status": "ok",
                    "result": execute_case(reference_case),
                }
            )
        except Exception as exc:  # noqa: BLE001 - engine failures are report data
            case_reports.append(
                {
                    "id": reference_case["id"],
                    "status": "engine_error",
                    "error": error_payload(exc),
                }
            )
    status = (
        "engine_error"
        if any(case_report["status"] == "engine_error" for case_report in case_reports)
        else "completed"
    )
    return {**_engine_header("python", corpus), "status": status, "cases": case_reports}


def _resolved_node_executable(node_executable: str | None) -> str:
    resolved = node_executable or shutil.which("node")
    if not resolved:
        raise RuntimeError("Node.js executable was not found.")
    return resolved


def run_typescript_corpus(
    corpus_path: Path,
    *,
    python_executable: str = sys.executable,
    node_executable: str | None = None,
) -> dict[str, Any]:
    completed = subprocess.run(
        [
            _resolved_node_executable(node_executable),
            str(TYPESCRIPT_BRIDGE),
            str(corpus_path.resolve()),
        ],
        cwd=ROOT,
        check=False,
        capture_output=True,
        text=True,
        encoding="utf-8",
        errors="replace",
        env={**os.environ, "MCA_CORPUS_PYTHON": python_executable},
    )
    if completed.returncode:
        detail = completed.stderr.strip() or completed.stdout.strip()
        raise RuntimeError(
            f"TypeScript corpus bridge failed with exit code {completed.returncode}: {detail}"
        )
    try:
        report = json.loads(completed.stdout)
    except json.JSONDecodeError as exc:
        raise RuntimeError("TypeScript corpus bridge returned invalid JSON.") from exc
    if not isinstance(report, dict) or report.get("engine") != "typescript":
        raise RuntimeError("TypeScript corpus bridge returned an invalid engine report.")
    return report


def fatal_engine_report(
    engine: str,
    corpus: dict[str, Any],
    error: BaseException,
) -> dict[str, Any]:
    return {
        **_engine_header(engine, corpus),
        "status": "engine_error",
        "error": error_payload(error),
        "cases": [],
    }

File: frontend/src/statisticalCorpusRunner.ts
import { createSeededSampleIndexDrawPort } from "./adapters/seededSampleIndexDrawPort";
import { createSimulationCommandFromNormalizedInput } from "./domain/simulation";
import type { SimulationResult } from "./domain/simulation";
import { createSimulationSeed } from "./domain/simulationValueObjects";
import { simulateMonteCarloLocal } from "./utils/simulation";

type CorpusInput = {
  throughput_samples: readonly number[];
  include_zero_weeks: boolean;
  mode: "backlog_to_weeks" | "weeks_to_items";
  backlog_size?: number;
  target_weeks?: number;
  n_sims: number;
};

export type ReferenceCase = {
  id: string;
  input: CorpusInput;
  seed: number;
};

export type StatisticalCorpus = {
  corpus_id: string;
  schema_version: string;
  normative_contract: {
    id: string;
    version: string;
  };
  prng_contract: { id: string };
  cases: readonly ReferenceCase[];
};

export type CanonicalResult = Record<string, unknown>;
export type TypeScriptCaseExecutor = (referenceCase: ReferenceCase) => CanonicalResult;

export type ValidationProbeDocument = {
  schema_version: string;
  normative_contract: string;
  cases: readonly {
    id: string;
    input: unknown;
    seed: unknown;
    accepted: boolean;
  }[];
};

export function canonicalizeTypeScriptResult(result: SimulationResult): CanonicalResult {
  const canonical: CanonicalResult = {
    result_kind: result.resultKind,
    result_percentiles: result.resultPercentiles,
  };
  if (result.riskScore !== undefined) {
    canonical.risk_score = result.riskScore;
  }
  canonical.result_distribution = result.resultDistribution;
  if (result.completionSummary !== undefined) {
    canonical.completion_summary = {
      completed_count: result.completionSummary.completedCount,
      censored_count: result.completionSummary.censoredCount,
      censored_rate: result.completionSummary.censoredRate,
      horizon_weeks: result.completionSummary.horizonWeeks,
    };
  }
  canonical.samples_count = result.samplesCount;
  if (result.throughputReliability === undefined) {
    throw new Error("throughput_reliability est requis dans la reponse canonique.");
  }
  canonical.throughput_reliability = {
    cv: result.throughputReliability.cv,
    iqr_ratio: result.throughputReliability.iqrRatio,
    slope_norm: result.throughputReliability.slopeNorm,
    label: result.throughputReliability.label,
    samples_count: result.throughputReliability.samplesCount,
  };
  canonical.seed = result.seed;
  return canonical;
}

export function executeTypeScriptCase(referenceCase: ReferenceCase): CanonicalResult {
  const command = createSimulationCommandFromNormalizedInput(
    referenceCase.input,
    createSimulationSeed(referenceCase.seed),
  );
  return canonicalizeTypeScriptResult(
    simulateMonteCarloLocal(
      command,
      createSeededSampleIndexDrawPort(command.seed),
    ),
  );
}

function errorPayload(error: unknown): { type: string; message: string } {
  if (error instanceof Error) {
    return { type: error.name, message: error.message };
  }
  return { type: typeof error, message: String(error) };
}

export function runTypeScriptCorpus(
  corpus: StatisticalCorpus,
  validationIssues: readonly string[],
  executeCase: TypeScriptCaseExecutor = executeTypeScriptCase,
): Record<string, unknown> {
  const header = {
    engine: "typescript",
    corpus_id: corpus.corpus_id,
    schema_version: corpus.schema_version,
    normative_contract: {
      id: corpus.normative_contract.id,
      version: corpus.normative_contract.version,
    },
    prng_contract: corpus.prng_contract.id,
  };
  if (validationIssues.length > 0) {
    return {
      ...header,
      status: "invalid_corpus",
      diagnostics: [...validationIssues],
      cases: [],
    };
  }
  const cases = corpus.cases.map((referenceCase) => {
    try {
      return {
        id: referenceCase.id,
        status: "ok",
        result: executeCase(referenceCase),
      };
    } catch (error) {
      return {
        id: referenceCase.id,
        status: "engine_error",
        error: errorPayload(error),
      };
    }
  });
  return {
    ...header,
    status: cases.some((caseReport) => caseReport.status === "engine_error")
      ? "engine_error"
      : "completed",
    cases,
  };
}

export function runTypeScriptValidationProbes(
  probes: ValidationProbeDocument,
): Record<string, unknown> {
  return {
    engine: "typescript",
    schema_version: probes.schema_version,
    normative_contract: probes.normative_contract,
    status: "completed",
    cases: probes.cases.map((probe) => {
      try {
        createSimulationCommandFromNormalizedInput(
          probe.input,
          createSimulationSeed(probe.seed),
        );
        return { id: probe.id, accepted: true };
      } catch {
        return { id: probe.id, accepted: false };
      }
    }),
  };
}
